feat(install): Add support for historical manifests by git history

Lutra-Fs opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!-- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

<!--
  By opening this PR you confirm that you have searched for similar issues/PRs here already.
  Failing to do so will most likely result in closing of this PR without any explanation.
  It is also mandatory to open a relevant issue for discussion with the maintainers,
  before creating any new PR.
  Read the contributing guide first to save both your and our time.
-->

#### Description
<!-- Describe your changes in detail -->

#### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Closes #3894 
<!-- or -->
Relates to #XXXX

#### How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

#### Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have ensured that I am targeting the `develop` branch.
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
- [x] I have added an entry in the CHANGELOG.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Install specific historical versions by searching local cache and bucket git history; manifest caching and cross-platform relative-path handling added. New setting use_git_history (default: true).

- **Improvements**
  - Prefer historical manifests before autoupdate; improved reliability of database queries, output handling, parameter binding, and resource cleanup; clearer failure messages when manifests can't be obtained.

- **Documentation**
  - Documented use_git_history and updated changelog.

- **Tests**
  - Added tests for history-based installs, relative paths, JSON parsing, and manifest-generation flows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Lutra-Fs committed:
feat(install): Add historical manifest search for app installation

- Implemented try-catch block in `install_app` function to handle installation failures and provide informative error messages.
- Added `Get-HistoricalManifest` function to search for specific app versions in git history, enhancing version management during installation.
- Updated `generate_user_manifest` to utilize historical manifests as a fallback when autoupdate fails.
- Enhanced documentation in `scoop-config.ps1` to clarify the use of git history for version installation.

Lutra-Fs commented:
One of the current implementation's issue is that when user provided a approximate version it will search all the git log entry which is a major performance bottleneck

Lutra-Fs committed:
docs(changelog): update changelog

niheaven commented:
The best way to search historical version is using sqlite feature enabled, and you should search the database first, if not found, use git history.

Lutra-Fs commented:
> The best way to search historical version is using sqlite feature enabled, and you should search the database first, if not found, use git history.

Will implement the SQLite version when I am available!

Lutra-Fs committed:
fix: Add Windows PowerShell compatibility for relative path operations

- Add Get-RelativePathCompat function that falls back to Uri-based implementation for PowerShell 5.x
- Replace direct GetRelativePath call with compatible wrapper
- Ensures historical manifest feature works across all PowerShell versions

Lutra-Fs committed:
style: Remove extra blank line from CHANGELOG.md

Lutra-Fs committed:
feat: Prioritize SQLite database for historical manifest search

- Add Get-HistoricalManifestFromDB function to search SQLite cache first
- Modify Get-HistoricalManifest to use SQLite before falling back to git history
- Implement best version matching for SQLite database results
- Add conditional database import to ensure functions are available
- Improve user messages to indicate whether cache or git history is used

This addresses @niheaven's suggestion to use SQLite for faster historical version searches, significantly reducing performance bottlenecks when searching for approximate versions.

Lutra-Fs committed:
fix: Remove trailing whitespaces from manifest.ps1

- Fix code style issues to pass tests
- Maintain existing functionality without changes

Lutra-Fs committed:
docs: Fix comment mismatch for python version example

- Change '3.7.x' to '3.12.x' to match the command 'python@3.12'
- Addresses GitHub Copilot's documentation consistency feedback

Lutra-Fs committed:
fix: Add PowerShell 5 compatibility for UTF8 encoding

- Use Out-UTF8File instead of Set-Content with UTF8Encoding for PowerShell 5
- Add version checks to use appropriate encoding methods
- Fixes 'Cannot bind parameter Encoding' error in Windows PowerShell 5
- Ensures historical manifest search works on all PowerShell versions

Lutra-Fs committed:
fix: Handle git command array output for all PowerShell versions

- Fix git show command returning arrays instead of strings
- Join array elements with newlines to create proper JSON content
- Remove misleading PowerShell 5 specific comments
- Fixes historical manifest search in both PowerShell 5 and 7

Lutra-Fs committed:
refactor: separate git history concerns from installation layer

- Remove git history fallback logic from install_app function
- Move git history concerns to manifest layer where they belong
- Improve separation of concerns and reduce coupling
- Update tests to reflect the cleaner design
- Merge git history tests into Scoop-Install.Tests.ps1 for consistency

This follows the 'Pull complexity downward' principle by keeping
installation logic focused on installation, while manifest resolution
handles all version fallback strategies internally.

Lutra-Fs committed:
revert: remove out-of-scope error handling changes from install.ps1

The try-catch block addition was a general error handling improvement
that belongs in a separate PR, not in this git history feature PR.

Lutra-Fs committed:
Fix test failures: improve parameter validation, file formatting, and test expectations

- Add missing psmodules.ps1 import to Install tests
- Fix architecture parameter validation (x64 -> 64bit)
- Add null/empty bucket parameter handling in Get-HistoricalManifestFromDB
- Fix file format issues (add missing newlines, remove trailing whitespace)
- Update JSON parsing tests to match actual implementation behavior
- Fix path separator normalization in Get-RelativePathCompat test
- Remove invalid error handling tests for non-existent try-catch blocks
- Reduce test failures from 11 to 4, with remaining issues being mock-related

All fixes respect existing codebase behavior and design decisions.

Lutra-Fs committed:
Remove best match logic from Get-HistoricalManifestFromDB

- Keep only exact match logic (lines 177-189)
- Return null immediately when no exact match found (line 193)
- Remove trailing whitespace from manifest.ps1
- Function now only supports exact version matches, no compatibility matching

Lutra-Fs committed:
fix: Update installation instructions to clarify search order for app versions

Lutra-Fs committed:
refactor: Enhance historical manifest retrieval with improved user guidance and fallback mechanisms

Lutra-Fs committed:
refactor: Remove deprecated install_app tests and enhance Get-HistoricalManifestFromDB for exact version matching

Lutra-Fs committed:
fix: Improve user guidance and error handling in generate_user_manifest function

Lutra-Fs committed:
refactor: Enhance Select-ScoopDBItem and Get-ScoopDBItem functions for improved data handling and output

Lutra-Fs committed:
refactor: remove new tests

Lutra-Fs commented:
> The best way to search historical version is using sqlite feature enabled, and you should search the database first, if not found, use git history.

Hi @niheaven 
I have implement the db source, now the order is
(if USE_GIT_HISTORY)
    (if USE_SQLITE_CACHE)
    SQLite Cache
    (else/failed)
    Git log --follow
(else/failed)
auto update with version variable

niheaven committed:
Merge branch 'develop' into feature/install_via_git_history

Lutra-Fs committed:
refactor(manifest): Simplify UTF-8 encoding handling and enhance test coverage

Lutra-Fs committed:
refactor(manifest): Streamline historical manifest retrieval and add Write-ManifestToUserCache helper function

Lutra-Fs commented on lib/manifest.ps1:
```diff
@@ -172,18 +172,301 @@ function Get-SupportedArchitecture($manifest, $architecture) {
+        warn "Error searching git history for '$app': $($_.Exception.Message)"
+        return $null
+    }
+}
```
This thing is kinda out of this PR's scope, and I believe it adds unnecessary complexity to the code

Lutra-Fs committed:
refactor(manifest): simplify history lookup and error handling; update tests

Lutra-Fs commented on test/Scoop-Manifest.Tests.ps1:
```diff
@@ -84,3 +94,145 @@ Describe 'Manifest Validator' -Tag 'Validator' {
+        Mock Write-Host {}
+        $p = generate_user_manifest 'foo' 'main' '1.0.0'
+        $p | Should -Be 'C:\git\foo.json'
+    }
```
has updated the behaviour to on autoupdate failure aborts with concise message, so this comment is outdated

Lutra-Fs commented on lib/manifest.ps1:
```diff
@@ -172,18 +172,301 @@ function Get-SupportedArchitecture($manifest, $architecture) {
+                info "Potential exact match '$versionFromMessage' found in commit message for $app (hash $hash). Verifying manifest..."
+                $manifestContent = Invoke-Git -Path $bucketDir -ArgumentList @('show', "$hash`:$relativeManifestPath")
+                if ($manifestContent -and ($LASTEXITCODE -eq 0)) {
+                    if ($manifestContent -is [Array]) {
```
Out of this PR's scope. mignt need to raise a new issue to fix that

Lutra-Fs replied:
Can you create one?

Lutra-Fs committed:
fix(manifest): enforce DataTable contract at caller; select latest matching version via h'^/h with validated fallback

Lutra-Fs committed:
test: add behavior-oriented mocks for Invoke-Git to simulate HEAD version retrieval

Lutra-Fs committed:
test: enhance Find-HistoricalManifestInCache with DataTable mock for improved manifest retrieval

niheaven committed:
Merge branch 'develop' into feature/install_via_git_history

niheaven committed:
Merge branch 'develop' into feature/install_via_git_history

niheaven committed:
Merge branch 'develop' into feature/install_via_git_history

Lutra-Fs commented on test/Scoop-Manifest.Tests.ps1:
```diff
@@ -84,3 +94,134 @@ Describe 'Manifest Validator' -Tag 'Validator' {
+        $result.version | Should -Be '1.2.3'
+        $result.ManifestText | Should -Match '"version":"1.2.3"'
+    }
+}
+
+Describe 'Find-HistoricalManifestInGit' -Tag 'Scoop' {
+    It 'returns $null when git history search disabled' {
+        Mock get_config -ParameterFilter { $name -eq 'use_git_history' } { $false }
+        $result = Find-HistoricalManifestInGit 'foo' 'main' '1.0.0'
+        $result | Should -Be $null
+    }
+
+    It 'returns manifest text on version match' {
+        $bucketRoot = 'C:\b\root'
+        $innerBucket = 'C:\b\root\bucket'
+        $umdir = Join-Path $env:TEMP 'ScoopTestsUM'
+        Mock get_config -ParameterFilter { $name -eq 'use_git_history' } { $true }
+        Mock Find-BucketDirectory -ParameterFilter { $Root } { $bucketRoot }
+        Mock Find-BucketDirectory -ParameterFilter { -not $Root } { $innerBucket }
+        Mock Test-Path -ParameterFilter { $Path -eq (Join-Path $bucketRoot '.git') } { $true }
+        Mock Test-Path -ParameterFilter { $Path -eq $innerBucket -and $PathType -eq 'Container' } { $true }
+        Mock Invoke-Git -ParameterFilter { $ArgumentList[0] -eq 'show' } { $global:LASTEXITCODE = 0; return '{"version":"1.0.0"}' }
+        Mock Invoke-Git -ParameterFilter { $ArgumentList[0] -eq 'log' } { @('abcdef0123456789') }
+
+        $result = Find-HistoricalManifestInGit 'foo' 'main' '1.0.0'
+        $result | Should -Not -BeNullOrEmpty
+        $result.version | Should -Be '1.0.0'
+        $result.ManifestText | Should -Match '"version":"1.0.0"'
+    }
+}
```
Adding two commits and asserting the chosen hash would add coverage for commit selection, but it's not guarding against a real bug we have - our selection loop isn't using HEAD or parent arbitrarily; it checks each candidate.

Lutra-Fs commented on lib/manifest.ps1:
```diff
@@ -172,38 +172,200 @@ function Get-SupportedArchitecture($manifest, $architecture) {
+
+    $dbResult = Get-ScoopDBItem -Name $app -Bucket $bucket -Version $requestedVersion
+
+    # Strictly follow DB contract: must be DataTable with at least one row
+    if (-not ($dbResult -is [System.Data.DataTable])) { return $null }
```
```suggestion
```

Lutra-Fs commented on lib/manifest.ps1:
```diff
@@ -172,38 +172,200 @@ function Get-SupportedArchitecture($manifest, $architecture) {
+            return $result
+        }
+    }
+
+    return $null
```
```suggestion
```
